Fix transitive dependencies not updating in PM UI

martinrrm opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/12333
Fixes: https://github.com/NuGet/Home/issues/12360
Fixes: https://github.com/NuGet/Home/issues/12428

Regression? Last working version: Regression introduced by https://github.com/NuGet/NuGet.Client/commit/e7ea3a13d3fc4738c375f2a48306d0cc46207e36

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->
Previously we tried to optimize this code path but introduced a regression. `IsInstalledAndTransitiveComputationNeeded` can tell us if a calculation of transitive dependencies is needed since its set to `True` when there is a change in assets file https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs#L386. 

This PR changes so that `IsInstalledAndTransitiveComputationNeeded` is true, we will always recalculate the transitive dependencies, if we don't do it, we will have an unsynchronized cache between transitive and installed dependencies and their transitive origins.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [x] Test exception - Already existing tests, adding manual testing videos for scenarios without tests
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

martinrrm committed:
fix transitive dependencies not updating in PM UI

martinrrm committed:
code improvemtns

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -134,14 +134,11 @@ internal async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool
             T installedPackages;
             T transitivePackages = default;
-            if (IsInstalledAndTransitiveComputationNeeded)
+            if (IsInstalledAndTransitiveComputationNeeded) // Cache invalidation
```
Does the comment add anything?

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -134,14 +134,11 @@ internal async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool
             {
                 // clear the transitive packages cache, since we don't know when a dependency has been removed
                 installedPackages = new T();
-                if (includeTransitivePackages)
```
Why remove this? It seems like it can be false no?

martinrrm replied:
If there's a change in the assets file `IsInstalledAndTransitiveComputationNeeded` is going to tell us and in the line below `await GetTargetsListAsync(assetsFilePath, token);` is going to recalculate the tagetsList which later will be used to calculate the transitive dependencies. 

I added `includeTransitivePackages` to the if above

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -186,7 +183,7 @@ internal async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool
             // get transitive packages
             IEnumerable<PackageReference> calculatedTransitivePackages = Enumerable.Empty<PackageReference>();
-            if (includeTransitivePackages)
+            if (includeTransitivePackages || IsInstalledAndTransitiveComputationNeeded) // Cache invalidation
```
Same as the previouis ones. The comment probably doesn't add much given that `IsInstalledAndTransitiveComputationNeeded` pretty much implies cache invalidation.

martinrrm committed:
pr comments

martinrrm committed:
pr comments

martinrrm commented:
Promoting transitive packages updates the counter and transitive packages list.
![transitive-dependencies-counter](https://user-images.githubusercontent.com/43253759/226453583-a1b4fd17-9ce1-435f-a4c7-326ff6c146ee.gif)

Removing a top-level package removes its transitive dependencies.
![transitive-dependencies-remove](https://user-images.githubusercontent.com/43253759/226453628-7778e9f7-7d84-4b21-be6d-c0ff8e04b7f1.gif)

jebriede commented:
Could you please add a description of what was changed to fix this in the Description section of the PR? Thanks!

jeffkl reviewed (dismissed):
Consider a unit test!

martinrrm committed:
fix installing multiple packages

martinrrm commented:
Added one test to check if the quantity of transitive packages is correct after multiple installations. Also fixed a bug where transitiveOrigins were not being updated because the if in https://github.com/NuGet/NuGet.Client/pull/5099/files#diff-cf69ebf676e61580349108d38e5dab1bbff91468194c6a8d342149d264139b15R201 wasn't correct.

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -186,7 +183,7 @@ internal async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool
             // get transitive packages
             IEnumerable<PackageReference> calculatedTransitivePackages = Enumerable.Empty<PackageReference>();
-            if (includeTransitivePackages)
+            if (includeTransitivePackages || IsInstalledAndTransitiveComputationNeeded)
```
This is the same condition as line 137 which had to be false for us to be in this else. Can you verify that we need to check both again and that the conditional logic is correct?

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -197,55 +194,52 @@ internal async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool
-                if (includeTransitiveOrigins && await ExperimentUtility.IsTransitiveOriginExpEnabled.GetValueAsync(token))
+                // Compute Transitive Origins
+                Dictionary<string, TransitiveEntry> transitiveOrigins;
+                if (includeTransitivePackages || IsInstalledAndTransitiveComputationNeeded
```
This is the same condition as line 137 which had to be false for us to be in this else. Can you verify that we need to check both again and that the conditional logic is correct?

jebriede reviewed (dismissed):
It would be great to have a test that verifies the behavior of this method when includeTransitivePackages is false but includeTransitiveOrigins is true and it's called for the first time. That would make sure that as we adjust the conditional logic, we don't break the unexpected case either. It should handle this in an expected way and gracefully.

If we have proper tests in place, I believe we can take a closer look at the places where includeTransitivePackages and IsInstalledAndTransitiveComputationNeeded are evaluated together with an OR and look if we need to check both in all places. What strikes me as concerning is how these bools are evaluated inside an else, where we can only be executing code if both the values were already false.

Cleaning up these extra checks would improve readability and maintainability. Approving with suggestions.

martinrrm committed:
added test

martinrrm committed:
improvements in the conditionals

martinrrm committed:
remove unncesary conditional

martinrrm committed:
fix test

martinrrm commented:
Closing, fix will be in https://github.com/NuGet/NuGet.Client/pull/5117
